### PR TITLE
[VM] Add Mutable and Immutable Field Borrow

### PR DESCRIPTION
**Pull request by tnowacki**
## Motivation

- Split `BorrowField` into `MutBorrowField` and `ImmBorrowField`
- `MutBorrowField` requires a mutable reference, with no borrows (mut or imm) in the tree
- `ImmBorrowField` requires a freezable reference (an imm ref or a mut ref with no mut borrows in the tree)
- This removes the sort of "dynamic" behavior in the borrow checker, where it would case on the type. The rules are technically the same in each case, regardless of the mut v. imm borrow (but in practice it cases on it for efficiency)
- I have considered doing this split twice in the past.
  - Last time, I stopped pushing for it since I convinced myself that a compiler's CSE pass could basically do the work for you; that is, You could fake having `ImmBorrowField` with `Freeeze` + `BorrowField` on a mutable reference, and then CSE would combine all the usages together to avoid conflicts. 
  - However! As I started thinking about how one would implement the CSE for this situation, a case popped up where it could not be done.  See the new tests for more details

- `BorrowLoc` will be split next
- We might want to split `BorrowGlobal`, but unlikely at this time (we will see)
  - Its less enticing due to additional limitations on mutable global references

## Test Plan

- New Tests
- cargo check; cargo test

### CODE REVIEW

File: language/vm/src/deserializer.rs [hunk 1]
```diff
@@ -1128,6 +1132,7 @@ impl Opcodes {
             0x33 => Ok(Opcodes::GET_TXN_SEQUENCE_NUMBER),
             0x34 => Ok(Opcodes::GET_TXN_PUBLIC_KEY),
             0x35 => Ok(Opcodes::FREEZE_REF),
+            0x36 => Ok(Opcodes::IMM_BORROW_FIELD),
```
Comment (sblackshear) [inline language/vm/src/deserializer.rs]: I hate to ask, but: shouldn't we reorder to make this closer to `MUT_BORROW_FIELD`? Totally happy with a TODO for this, but seems like something we should do eventually.

File: language/vm/src/file_format_common.rs [hunk 2]
```diff
@@ -170,6 +170,7 @@ pub enum Opcodes {
     GET_TXN_SEQUENCE_NUMBER = 0x33,
     GET_TXN_PUBLIC_KEY      = 0x34,
     FREEZE_REF              = 0x35,
+    IMM_BORROW_FIELD        = 0x36,
```
Comment (sblackshear) [inline language/vm/src/file_format_common.rs]: Similar

File: language/bytecode_verifier/src/type_memory_safety.rs [hunk 3]
```diff
@@ -221,63 +221,105 @@ impl<'a> TypeAndMemorySafetyAnalysis<'a> {
...
+                    state.borrowed_nonces_for_field(*field_definition_index, operand_nonce.clone());
+                if !Self::write_borrow_ok(borrowed_nonces) {
+                    // TODO Separate Mut and Imm BorrowField Errors
+                    return Err(VMStaticViolation::BorrowFieldExistsMutableBorrowError(
+                        offset,
+                    ));
+                }
+
+                self.stack.push(StackAbstractValue {
+                    signature: SignatureToken::MutableReference(Box::new(field_signature.clone())),
```
Comment (sblackshear) [inline language/bytecode_verifier/src/type_memory_safety.rs]: `clone` needed?

File: language/bytecode_verifier/src/type_memory_safety.rs [hunk 4]
```diff
@@ -221,63 +221,105 @@ impl<'a> TypeAndMemorySafetyAnalysis<'a> {
...
+                    // TODO Separate Mut and Imm BorrowField Errors
+                    return Err(VMStaticViolation::BorrowFieldBadFieldError(offset));
+                }
+
+                let field_signature = self
+                    .module
+                    .get_field_signature(*field_definition_index)
+                    .0
+                    .clone();
+                let operand_nonce = operand.value.extract_nonce().unwrap().clone();
```
Comment (sblackshear) [inline language/bytecode_verifier/src/type_memory_safety.rs]: `clone` needed?

File: language/compiler/ir_to_bytecode/src/compiler.rs [hunk 5]
```diff
@@ -2196,22 +2196,14 @@ impl<S: Scope + Sized> Compiler<S> {
...
-        Ok(if is_mutable {
-            if !input_is_mutable {
-                bail!("Unsupported Syntax: Cannot take a mutable field reference in an immutable reference. It is not expressible in the bytecode");
-            }
+        Ok(if is_mutable_borrow {
+            code.code.push(Bytecode::MutBorrowField(fd_idx));
+            function_frame.push()?;
             self.make_singleton_vec_deque(InferredType::MutableReference(inner_token))
         } else {
-            if input_is_mutable {
```
Comment (sblackshear) [inline language/compiler/ir_to_bytecode/src/compiler.rs]: Just looking at how much simpler this code gets seems to suggest that this is the right decision...

File: language/vm/src/file_format_common.rs (hunk 2 above)
Comment (tnowacki) [reply language/vm/src/file_format_common.rs]: Done

Comment (tnowacki) [reply language/vm/src/file_format_common.rs]: Done

File: language/bytecode_verifier/src/type_memory_safety.rs (hunk 4 above)
Comment (tnowacki) [reply language/bytecode_verifier/src/type_memory_safety.rs]: I think all of these clones are due to references being used

Comment (tnowacki) [reply language/bytecode_verifier/src/type_memory_safety.rs]: Ah, nevermind. the operand one is necessary, but not the signature one

**Pull request merged**